net: l2: bluetooth: Fix IPSP interface startup at boot

rlubos opened the pull request:
Bluetooth IPSP interface failed to be brought up at boot due to some invalid checks (overlooked in https://github.com/zephyrproject-rtos/zephyr/pull/51240). This PR fixes it.

rlubos committed:
net: l2: bluetooth: Remove invalid check

After network interfaces state handling rework, it's no longer correct
to verify the connection status on `net_if_up()`, as this only changes
the administrative state of the interface. The interface won't be put in
operational UP state until it's connected.

This check prevented the interface from being brought up during system
boot.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: if: Allow to set LL address after interface was brough admin UP

Bluetooth IPSP L2 sets the LL address only after establishing Bluetooth
connection. As this can take place after the interface was brough UP by
the application, the network interface API should not block such
attempt. Instead, verify the NET_IF_RUNNING flag.

For the same reason, move the assert verifying that LL address is set
from net_if_up() (which only puts the interface in the admin UP state)
into the function that transitions the interface into operational UP
state.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jori-nordic reviewed (dismissed):
Changes look ok, even if I don't fully understand them.

jori-nordic commented:
@rlubos would https://github.com/zephyrproject-rtos/zephyr/issues/50345 also be fixed by this patch?

rlubos commented:
> @rlubos would #50345 also be fixed by this patch?

From the issue description, I don't think it will. The IPv6 stack will send its notifications as soon as it gets notified by the Bluetooth L2 it's connected, nothing changed here.

jukkar commented on include/zephyr/net/net_if.h:
```diff
@@ -857,10 +857,6 @@ static inline int net_if_set_link_addr_unlocked(struct net_if *iface,
 						uint8_t *addr, uint8_t len,
 						enum net_link_type type)
 {
-	if (net_if_flag_is_set(iface, NET_IF_UP)) {
-		return -EPERM;
-	}
-
```
I am not sure about this. Don't we need some check here so that it is not possible to set the address if the interface is up and running as that would confuse the IPv6 stack? Perhaps something like
```
	if (net_if_flag_is_set(iface, NET_IF_RUNNING)) {
		return -EPERM;
	}
```
WDYT?

rlubos replied:
I need to investigate 802.15.4 a bit, I've tried to leave the check as suggested, but there's some hacky code in 802.15.4 L2 causing some tests to fail:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/l2/ieee802154/ieee802154_mgmt.c#L206

Worst case I'll need to update the flag in 802.15.4 L2 as well..

rlubos replied:
It seems that there is a case for IEEE802154 when the LL address can be updated during interface operation (processing MAC command, updating to short address). Therefore for now, I've updated the workaround which allowed to do this by using `NET_IF_RUNNING` flag instead of `NET_IF_UP`.

rlubos committed:
net: l2: ieee802154: Align LL address update routine

IEEE802154 L2 may modify the LL address during interface operation (when
processing MAC command). So far the L2 workaround the LL address update
protection by clearing the NET_IF_UP flag temporarily, but due to recent
changes it no longer works. Update this workaround to verify
NET_IF_RUNNING flag instead.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar reviewed (approved):
LGTM
